In [3]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
!unzip '/content/drive/MyDrive/MapsDataCities.zip'

In [ ]:
import os
from PIL import Image
f = r'/content/drive/MyDrive/MapsDataCities'
#f_open = r'/content/drive/MyDrive/maps_data/patterns'
#f_save = r'/content/drive/MyDrive/maps_data'
#img_number = 0

for file in os.listdir(f):
  #print(file)
  try:
    #print(re.search(r'\d+', file).group(0))
    #print(file)
    f_img = f +"/" + file
    #f_number = f + "/" + re.search(r'\d+', file).group(0)
    #os.rename(f_img, str(f_number))
    image = Image.open(f_img)
  except:
    print("//")
    print("filename no number")
    print(file)
    print("//")
  #img_number += 1
  #f_img = f_open+"/"+file
  #f_img_save = f_save + "/" + file
  #image = Image.open(f_img)
  #print(image.size)
#  try:
#    f_img = f+"/"+file
#    image = Image.open(f_img)
#    image = image.resize((301,301),Image.ANTIALIAS)
#    image.save(f_img_save,quality=95)
#    os.rename(f_img_save, str(f_img_save))
#  except:
#    print("exception, shame")

In [4]:
import pathlib
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import matplotlib.pyplot as plt
from torch.utils.data.sampler import SubsetRandomSampler
from torch.utils.data import DataLoader
from torchvision import datasets, transforms
import matplotlib.pyplot as plt
%matplotlib inline
import torch.nn as nn
import tensorflow as tf

In [5]:
kernel_size = 4
stride = 2
padding = 4
dilation= 2
init_kernel = 8 # initial number of filters

In [6]:
import os

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.realpath('__file__')))
print(PROJECT_PATH)
#DATAPATH = os.path.join(PROJECT_PATH,'content/drive/MyDrive','maps_data')
DATAPATH = os.path.join(PROJECT_PATH,'content/drive/MyDrive','MapsDataCities')
print(DATAPATH)

/
/content/drive/MyDrive/MapsDataCities


In [7]:
data_root = pathlib.Path(DATAPATH)
print(data_root)

/content/drive/MyDrive/MapsDataCities


In [8]:
import sys
import os
import tensorflow as tf
import random


def preprocess_image(image):
    image = tf.io.decode_png(image, channels=3)
    image = tf.image.resize(image, [286, 286])
    image /= 255.0  # normalize to [0,1] range

    return image


def load_and_preprocess_image(path):
    image = tf.io.read_file(path)
    return preprocess_image(image)


def get_filepaths(dirpath):
    """
    Return paths of all files in a directory.
    :param dirpath:
    :return:
    """
    data_root = pathlib.Path(dirpath)
    all_image_paths = list(data_root.glob('*'))
    all_image_paths = [str(path) for path in all_image_paths]
    random.shuffle(all_image_paths)

    return all_image_paths


def load_dataset(dirpath):

    all_image_paths = get_filepaths(dirpath)
    path_ds = tf.data.Dataset.from_tensor_slices(all_image_paths)
    image_ds = path_ds.map(load_and_preprocess_image, num_parallel_calls=tf.data.experimental.AUTOTUNE)

    return image_ds

In [9]:
ds = load_dataset(data_root)

In [10]:
import numpy as np
import torchvision

In [11]:
# define a Conv VAE
class ConvVAE(nn.Module):
    def __init__(self):
        super(ConvVAE, self).__init__()
 
        # encoder
        self.enc1 = nn.Conv2d(
            in_channels=3, out_channels=init_kernel, kernel_size=kernel_size, 
            stride=stride, padding=padding
        )
        self.enc2 = nn.Conv2d(
            in_channels=init_kernel, out_channels=init_kernel*2, kernel_size=kernel_size, 
            stride=stride#, padding=padding
        )
        self.enc3 = nn.Conv2d(
            in_channels=init_kernel*2, out_channels=init_kernel*4, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        self.enc4 = nn.Conv2d(
            in_channels=init_kernel*4, out_channels=init_kernel*8, kernel_size=kernel_size, 
            stride=stride#, padding=padding
        )
        self.enc5 = nn.Conv2d(
            in_channels=init_kernel*8, out_channels=init_kernel, kernel_size=kernel_size, 
            stride=stride#, padding=padding
        )
        # decoder 
        self.dec1 = nn.ConvTranspose2d(
            in_channels=init_kernel, out_channels=init_kernel*8, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        self.dec2 = nn.ConvTranspose2d(
            in_channels=init_kernel*8, out_channels=init_kernel*4, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        self.dec3 = nn.ConvTranspose2d(
            in_channels=init_kernel*4, out_channels=init_kernel*2, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        self.dec4 = nn.ConvTranspose2d(
            in_channels=init_kernel*2, out_channels=init_kernel, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        self.dec5 = nn.ConvTranspose2d(
            in_channels=init_kernel, out_channels=3, kernel_size=kernel_size, 
            stride=stride#, padding=padding, dilation=dilation
        )
        
    def encode(self, x):
        print(x.shape)
        x = F.relu(self.enc1(x))
        x = F.relu(self.enc2(x))
        x = F.relu(self.enc3(x))
        x = F.relu(self.enc4(x))
        x = self.enc5(x)
        mu = x
        log_var = x

        # get the latent vector through reparameterization
        z = self.reparameterize(mu, log_var)
        return z, mu, log_var
    
    def decode(self, z):
        x = F.relu(self.dec1(z))
        #print(x.shape)
        x = F.relu(self.dec2(x))
        #print(x.shape)
        x = F.relu(self.dec3(x))
        #print(x.shape)
        x = F.relu(self.dec4(x))
        #print(x.shape)
        return torch.sigmoid(self.dec5(x))

    def reparameterize(self, mu, log_var):
        """
        :param mu: mean from the encoder's latent space
        :param log_var: log variance from the encoder's latent space
        """
        std = torch.exp(0.5*log_var) # standard deviation
        eps = torch.randn_like(std) # `randn_like` as we need the same size
        sample = mu + (eps * std) # sampling
        return sample
 
    def forward(self, x):
        # encoding
        x = F.relu(self.enc1(x))
        #print("enc1: " + str(x.size()))
        x = F.relu(self.enc2(x))
        #print("enc2: " + str(x.size()))
        x = F.relu(self.enc3(x))
        #print("enc3: " + str(x.size()))
        x = F.relu(self.enc4(x))
        #print("enc4: " + str(x.size()))
        x = self.enc5(x)
        #print("enc5: " + str(x.size()))
        # get `mu` and `log_var`
        mu = x
        log_var = x
        # get the latent vector through reparameterization
        z = self.reparameterize(mu, log_var)
        #print(z.shape)
 
        # decoding
        x = F.relu(self.dec1(z))
        #print("dec1: " + str(x.size()))
        #print("In channels: " + str(init_kernel))
        #print("Out channels: " + str(init_kernel*8))
        #print("Kernel Size:  constant at 10")
        x = F.relu(self.dec2(x))
        #print("dec2: " + str(x.size()))
        #print("In channels: " + str(init_kernel*8))
        #print("Out channels: " + str(init_kernel*4))
        #print("Kernel Size:  constant at 10")
        x = F.relu(self.dec3(x))
        #print("dec3: " + str(x.size()))
        #print("In channels: " + str(init_kernel*4))
        #print("Out channels: " + str(init_kernel*2))
        #print("Kernel Size:  constant at 10")
        x = F.relu(self.dec4(x))
        #print("dec4: " + str(x.size()))
        #print("In channels: " + str(init_kernel*2))
        #print("Out channels: " + str(init_kernel))
        #print("Kernel Size:  constant at 10")
        reconstruction = torch.sigmoid(self.dec5(x))
        #print("reconstruction: " + str(reconstruction.size()))
        #print("In channels: " + str(init_kernel))
        #print("Out channels: " + str(3))
        #print("Kernel Size:  constant at 10")
        return reconstruction, mu, log_var, z

In [12]:
import scipy.io
import torch
import torch.optim as optim
import torch.nn as nn
#import model
from torch.utils.data import DataLoader, Dataset
from tqdm import tqdm
from torchvision.utils import save_image

In [13]:
# learning parameters
batch_size = 128
#lr = 0.001
lr = 0.001
epochs = 900
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

In [14]:
from torch.utils import data
import sys

#sys.path.append('../')
from torchvision import datasets
# from skimage import io, transform
import os
import re

from PIL import Image

import torch
from PIL import Image, ImageOps
import numpy as np

import torchvision
import torchvision.transforms.functional as trF


class Invert(object):
    """Inverts the color channels of an PIL Image
    while leaving intact the alpha channel.
    """

    def invert(self, img):
        r"""Invert the input PIL Image.
        Args:
            img (PIL Image): Image to be inverted.
        Returns:
            PIL Image: Inverted image.
        """
        #if not trF._is_pil_image(img):
        #    raise TypeError('img should be PIL Image. Got {}'.format(type(img)))

        if img.mode == 'RGBA':
            r, g, b, a = img.split()
            inv = Image.merge('RGB', (r, g, b))
            #inv = ImageOps.invert(rgb)
            #r, g, b = inv.split()
            #inv = Image.merge('RGBA', (r, g, b, a))
        elif img.mode == 'LA':
            l, a = img.split()
            l = ImageOps.invert(l)
            inv = Image.merge('LA', (l, a))
        else:
            inv = ImageOps.invert(img)
        return inv

    def __call__(self, img):
        """
        Args:
            img (PIL Image): Image to be inverted.
        Returns:
            PIL Image: Inverted image.
        """
        return self.invert(img)

    def __repr__(self):
        return self.__class__.__name__ + '()'

class CityImageDataset(data.Dataset):
    """
    Urban images dataset.
    """

    def __init__(self, root, transform=None):
        self.root = root
        self.transform = transform
        self.image_files = [f for f in os.listdir(root) if not f.startswith('.')]
        #self.image_names = [''.join(re.findall(r'[^\/]+(?=\.)',f)) for f in os.listdir(root) if not f.startswith('.')]

    def __len__(self):
        return len(self.image_files)

    def __getitem__(self, idx):
        img_name = os.path.join(self.root,self.image_files[idx])
        sample = Image.open(img_name)

        if self.transform:
            sample = self.transform(sample)

        return sample

    def __getname__(self, idx):
        return self.image_files[idx]

In [15]:
to_pil_image = transforms.ToPILImage()

city_dataset = CityImageDataset(root=data_root,transform=torchvision.transforms.Compose([    
    #torchvision.transforms.Grayscale(num_output_channels=3),                                                                                 
    Invert(),
    torchvision.transforms.RandomCrop(286),
    torchvision.transforms.ToTensor(),
    #transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
]))

In [16]:
validation_split = .2
shuffle_dataset = True
random_seed= 42
dataset_size = len(city_dataset)
indices = list(range(dataset_size))
split = int(np.floor(validation_split * dataset_size))
if shuffle_dataset :
    np.random.seed(random_seed)
    np.random.shuffle(indices)
train_indices, val_indices = indices[split:], indices[:split]

In [ ]:
print(len(train_indices))
print(len(val_indices))
print(dataset_size)

12050
3012
15062


In [17]:
from torch.utils.data.sampler import SubsetRandomSampler

train_sampler = SubsetRandomSampler(train_indices)
valid_sampler = SubsetRandomSampler(val_indices)

train_loader = torch.utils.data.DataLoader(city_dataset, batch_size=batch_size, 
                                           sampler=train_sampler)
validation_loader = torch.utils.data.DataLoader(city_dataset, batch_size=batch_size,
                                                sampler=valid_sampler)

In [ ]:
#print(''.join(re.findall(r'[^\/]+(?=\.)',train_loader.dataset.__getname__(2))))
print(train_loader.dataset.__getname__(2))

19734.png


In [18]:
#Pytorch Autoencoder
from torchvision.utils import save_image
from torchvision.utils import make_grid
import imageio
def get_device():
    if torch.cuda.is_available():
        print("gpu")
        device = 'cuda:0'
    else:
        print("cpu")
        device = 'cpu'
    return device
def make_dir():
    image_dir = 'CityDatasetMap_Images'
    if not os.path.exists(image_dir):
        os.makedirs(image_dir)
def save_decoded_image(img, name):
    img = img.view(img.size(0), 3, 286, 286)
    save_image(img, name)
def make_dir():
    image_dir = 'CityDatasetMap_Images'
    if not os.path.exists(image_dir):
        os.makedirs(image_dir)
def image_to_vid(images):
    imgs = [np.array(to_pil_image(img)) for img in images]
    imageio.mimsave('./CityDatasetMap_Images/generated_images.gif', imgs)
def save_reconstructed_images(recon_images, epoch):
    save_image(recon_images.cpu(), f"./CityDatasetMap_Images/output{epoch}.jpg")
def save_loss_plot(train_loss, valid_loss):
    plt.figure(figsize=(10, 7))
    plt.plot(train_loss, color='orange', label='train loss')
    plt.plot(valid_loss, color='red', label='validataion loss')
    plt.xlabel('Epochs')
    plt.ylabel('Loss')
    plt.legend()
    plt.savefig('./CityDatasetMap_Images/loss_graph.jpg')
    plt.show()

In [19]:
model = ConvVAE().to(device)
optimizer = optim.Adam(model.parameters(), lr=lr)
criterion = nn.BCELoss(reduction='sum')

In [20]:
def final_loss(bce_loss, mu, logvar, beta):
    """
    This function will add the reconstruction loss (BCELoss) and the 
    KL-Divergence.
    KL-Divergence = 0.5 * sum(1 + log(sigma^2) - mu^2 - sigma^2)
    :param bce_loss: recontruction loss
    :param mu: the mean from the latent vector
    :param logvar: log variance from the latent vector
    """
    BCE = bce_loss 
    KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
    return BCE + 1 + beta * KLD

In [21]:
def fit(model, dataloader, beta):
    model.train()
    running_loss = 0.0
    for i, data in tqdm(enumerate(dataloader), total=int(len(train_indices)/dataloader.batch_size)):
        #print("data " + str(i))
        #print(dataloader.dataset.__getname__(i))
        data = data
        data = data.to(device)
        data = data
        optimizer.zero_grad()
        reconstruction, mu, logvar, z = model(data)
        #print(reconstruction.size())
        #print(data.size())
        bce_loss = criterion(reconstruction, data)
        loss = final_loss(bce_loss, mu, logvar, beta)
        loss.backward()
        running_loss += loss.item()
        optimizer.step()
    train_loss = running_loss/len(dataloader.dataset)
    return train_loss

In [22]:
def validate(model, dataloader, beta):
    model.eval()
    running_loss = 0.0
    with torch.no_grad():
        for i, data in tqdm(enumerate(dataloader), total=int(len(val_indices)/dataloader.batch_size)):
            data = data
            data = data.to(device)
            data = data
            reconstruction, mu, logvar, z = model(data)
            bce_loss = criterion(reconstruction, data)
            loss = final_loss(bce_loss, mu, logvar, beta)
            running_loss += loss.item()
        
            # save the last batch input and output of every epoch
            if i == int(len(val_indices)/dataloader.batch_size) - 1:
                num_rows = 8
                both = torch.cat((data[:8], reconstruction[:8]))
                recon_images = reconstruction
                #save_image(both.cpu(), f"../outputs/output{epoch}.png", nrow=num_rows)
                save_decoded_image(both.cpu().data, name='./CityDatasetMap_Images/decoded{}.png'.format(epoch))
    val_loss = running_loss/len(dataloader.dataset)
    return val_loss #, recon_images

In [ ]:
train_loss = []
val_loss = []
#grid_images = []
make_dir() 
beta = 6.5
min_beta = 1
for epoch in range(epochs):
    print(f"Epoch {epoch+1} of {epochs}")
    #print(train_loader)
    train_epoch_loss = fit(model, train_loader, beta)
    #val_epoch_loss, recon_images = validate(model, validation_loader, beta)
    val_epoch_loss = validate(model, validation_loader, beta)
    beta *= 0.998
    if beta < min_beta: 
        beta = min_beta
    train_loss.append(train_epoch_loss)
    val_loss.append(val_epoch_loss)
    print(f"Train Loss: {train_epoch_loss:.4f}")
    print(f"Val Loss: {val_epoch_loss:.4f}")
    # save the reconstructed images from the validation loop
    #save_reconstructed_images(recon_images, epoch+1)
    # convert the reconstructed images to PyTorch image grid format
    #image_grid = make_grid(recon_images.detach().cpu())
    #grid_images.append(image_grid)

# save the reconstructions as a .gif file
#image_to_vid(grid_images)
# save the loss plots to disk
save_loss_plot(train_loss, val_loss)
print('TRAINING COMPLETE')

  0%|          | 0/94 [00:00<?, ?it/s]

Epoch 1 of 900


95it [1:12:03, 45.51s/it]
24it [17:15, 43.16s/it]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 99655.4841
Val Loss: 20165.9882
Epoch 2 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 78300.2832
Val Loss: 18948.1465
Epoch 3 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 75704.9566
Val Loss: 18549.5805
Epoch 4 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 74679.8323
Val Loss: 18407.9631
Epoch 5 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 74129.7536
Val Loss: 18267.1507
Epoch 6 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 73717.1165
Val Loss: 18255.0232
Epoch 7 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 73504.9771
Val Loss: 18142.3315
Epoch 8 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 73291.2085
Val Loss: 18105.4121
Epoch 9 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 73097.6556
Val Loss: 18170.2537
Epoch 10 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72982.1692
Val Loss: 18082.3041
Epoch 11 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72756.5417
Val Loss: 18113.1929
Epoch 12 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72638.8582
Val Loss: 18010.6396
Epoch 13 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72485.9933
Val Loss: 17885.9216
Epoch 14 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72258.0668
Val Loss: 17886.3879
Epoch 15 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72129.6390
Val Loss: 17865.6818
Epoch 16 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71950.8561
Val Loss: 17772.9093
Epoch 17 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71891.6639
Val Loss: 17748.2219
Epoch 18 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71637.0655
Val Loss: 17741.7119
Epoch 19 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71469.7187
Val Loss: 18015.7741
Epoch 20 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 72207.7534
Val Loss: 17696.9423
Epoch 21 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71226.1920
Val Loss: 17603.6402
Epoch 22 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 71054.7516
Val Loss: 17560.2092
Epoch 23 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70902.3146
Val Loss: 17541.7370
Epoch 24 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70773.7175
Val Loss: 17491.6309
Epoch 25 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70653.1132
Val Loss: 17473.9756
Epoch 26 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70563.1008
Val Loss: 17458.0059
Epoch 27 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70506.1233
Val Loss: 17414.7076
Epoch 28 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70327.9760
Val Loss: 17389.8083
Epoch 29 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70189.3744
Val Loss: 17374.4934
Epoch 30 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70131.3374
Val Loss: 17620.1576
Epoch 31 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 70176.3003
Val Loss: 17395.9551
Epoch 32 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69942.2499
Val Loss: 17292.0579
Epoch 33 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69836.7569
Val Loss: 17301.2528
Epoch 34 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69777.2742
Val Loss: 17389.5755
Epoch 35 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69838.1852
Val Loss: 17259.6920
Epoch 36 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69622.7434
Val Loss: 17371.0239
Epoch 37 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69647.5998
Val Loss: 17207.2264
Epoch 38 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69470.0700
Val Loss: 17192.9690
Epoch 39 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69384.4649
Val Loss: 17205.0889
Epoch 40 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69381.3872
Val Loss: 17178.3026
Epoch 41 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69285.0474
Val Loss: 17202.1161
Epoch 42 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69248.3985
Val Loss: 17139.3115
Epoch 43 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69168.7264
Val Loss: 17248.3610
Epoch 44 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69461.5087
Val Loss: 17108.7250
Epoch 45 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69048.1534
Val Loss: 17101.8753
Epoch 46 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69017.3026
Val Loss: 17117.3976
Epoch 47 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69026.8814
Val Loss: 17277.8878
Epoch 48 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69112.7424
Val Loss: 17060.2946
Epoch 49 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68854.3648
Val Loss: 17042.1830
Epoch 50 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68861.3942
Val Loss: 17118.5426
Epoch 51 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68848.5654
Val Loss: 17108.7537
Epoch 52 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68802.6497
Val Loss: 17014.6325
Epoch 53 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68692.2728
Val Loss: 17226.5136
Epoch 54 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68838.0238
Val Loss: 17101.2473
Epoch 55 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68695.5691
Val Loss: 17021.1278
Epoch 56 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68659.9501
Val Loss: 16980.9824
Epoch 57 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68540.0005
Val Loss: 16985.1033
Epoch 58 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68509.7913
Val Loss: 17039.9187
Epoch 59 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68509.8379
Val Loss: 16991.0048
Epoch 60 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68496.5442
Val Loss: 17571.7416
Epoch 61 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69799.2843
Val Loss: 16966.4889
Epoch 62 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68434.3074
Val Loss: 16962.4777
Epoch 63 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68383.7123
Val Loss: 16943.9272
Epoch 64 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68326.8114
Val Loss: 16922.2953
Epoch 65 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68326.3571
Val Loss: 16907.3981
Epoch 66 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68255.9630
Val Loss: 16950.6159
Epoch 67 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68250.6707
Val Loss: 16894.9324
Epoch 68 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68211.1344
Val Loss: 17033.9776
Epoch 69 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68579.2573
Val Loss: 16886.4617
Epoch 70 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68137.5343
Val Loss: 16868.4708
Epoch 71 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68107.3345
Val Loss: 16903.3582
Epoch 72 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68134.2735
Val Loss: 16856.8251
Epoch 73 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68078.1274
Val Loss: 16859.7948
Epoch 74 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68027.2458
Val Loss: 16974.8641
Epoch 75 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68174.6363
Val Loss: 16834.8556
Epoch 76 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67963.2528
Val Loss: 16865.8548
Epoch 77 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67957.1932
Val Loss: 16866.2709
Epoch 78 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67919.6220
Val Loss: 16825.8840
Epoch 79 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67916.3332
Val Loss: 17106.3146
Epoch 80 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 69633.0910
Val Loss: 17028.1478
Epoch 81 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68116.1096
Val Loss: 16906.0699
Epoch 82 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67973.9912
Val Loss: 16834.0627
Epoch 83 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67876.0085
Val Loss: 16816.3217
Epoch 84 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67831.5497
Val Loss: 16803.3936
Epoch 85 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67852.9598
Val Loss: 16816.3805
Epoch 86 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67799.4001
Val Loss: 16781.7204
Epoch 87 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67762.7557
Val Loss: 16787.0946
Epoch 88 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67758.3092
Val Loss: 16962.9151
Epoch 89 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67859.2312
Val Loss: 16769.7500
Epoch 90 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67689.7711
Val Loss: 16760.0103
Epoch 91 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67672.6122
Val Loss: 16906.2198
Epoch 92 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 68097.7383
Val Loss: 16789.3468
Epoch 93 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67642.6768
Val Loss: 16753.6169
Epoch 94 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67613.6598
Val Loss: 16768.8945
Epoch 95 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67679.9923
Val Loss: 16764.2955
Epoch 96 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67569.9691
Val Loss: 16739.0878
Epoch 97 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67591.5284
Val Loss: 16736.3869
Epoch 98 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67547.7193
Val Loss: 16782.9207
Epoch 99 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67866.7470
Val Loss: 16725.8959
Epoch 100 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67516.2352
Val Loss: 16729.5562
Epoch 101 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67511.5448
Val Loss: 16764.5375
Epoch 102 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67502.7447
Val Loss: 17008.2407
Epoch 103 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67749.8002
Val Loss: 16739.9666
Epoch 104 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67452.6747
Val Loss: 16702.9796
Epoch 105 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67441.8953
Val Loss: 16702.6336
Epoch 106 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67392.9806
Val Loss: 16699.3011
Epoch 107 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67451.6373
Val Loss: 16694.3346
Epoch 108 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67394.3603
Val Loss: 16867.7833
Epoch 109 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67783.1289
Val Loss: 16687.5063
Epoch 110 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67369.6935
Val Loss: 16680.1745
Epoch 111 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67346.4189
Val Loss: 16694.0622
Epoch 112 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67360.1940
Val Loss: 16675.4799
Epoch 113 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67324.5612
Val Loss: 16674.6780
Epoch 114 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67348.6113
Val Loss: 16670.7638
Epoch 115 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67328.3125
Val Loss: 16676.2608
Epoch 116 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67292.5627
Val Loss: 16671.8505
Epoch 117 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67380.0002
Val Loss: 16659.3248
Epoch 118 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67261.3916
Val Loss: 16702.3688
Epoch 119 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67415.3941
Val Loss: 16928.1094
Epoch 120 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67301.1953
Val Loss: 16653.9991
Epoch 121 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67210.3125
Val Loss: 16746.8015
Epoch 122 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67253.7789
Val Loss: 16840.4813
Epoch 123 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67359.2829
Val Loss: 16698.0854
Epoch 124 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67205.5550
Val Loss: 16666.9186
Epoch 125 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67190.1943
Val Loss: 16666.0819
Epoch 126 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67231.4977
Val Loss: 16636.3428
Epoch 127 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67144.2478
Val Loss: 16662.8749
Epoch 128 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67168.1128
Val Loss: 16660.4210
Epoch 129 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67149.1801
Val Loss: 16628.7130
Epoch 130 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67145.1673
Val Loss: 16619.9658
Epoch 131 of 900


95it [01:05,  1.46it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67129.4077
Val Loss: 16623.3796
Epoch 132 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67324.5006
Val Loss: 16620.2912
Epoch 133 of 900


95it [01:05,  1.46it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67082.9415
Val Loss: 16702.5400
Epoch 134 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67138.4309
Val Loss: 16704.9645
Epoch 135 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67199.3815
Val Loss: 16732.0477
Epoch 136 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67070.5475
Val Loss: 16610.4675
Epoch 137 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67021.5930
Val Loss: 16629.6335
Epoch 138 of 900


95it [01:05,  1.44it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67022.6613
Val Loss: 16602.3417
Epoch 139 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67016.4503
Val Loss: 16588.3899
Epoch 140 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.52it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67058.4296
Val Loss: 16774.3689
Epoch 141 of 900


95it [01:04,  1.46it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67839.0992
Val Loss: 16620.3018
Epoch 142 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67007.2311
Val Loss: 16586.4376
Epoch 143 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66958.0336
Val Loss: 16588.2426
Epoch 144 of 900


95it [01:04,  1.47it/s]
24it [00:15,  1.52it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66971.9624
Val Loss: 16576.4018
Epoch 145 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66934.1508
Val Loss: 16602.4706
Epoch 146 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66943.6959
Val Loss: 16577.1715
Epoch 147 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66933.2905
Val Loss: 16587.8987
Epoch 148 of 900


95it [01:05,  1.46it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66977.0637
Val Loss: 16605.9277
Epoch 149 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.52it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66942.3682
Val Loss: 16584.3068
Epoch 150 of 900


95it [01:05,  1.46it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66875.7796
Val Loss: 16589.5241
Epoch 151 of 900


95it [01:04,  1.46it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66971.6173
Val Loss: 16566.0181
Epoch 152 of 900


95it [01:05,  1.46it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66910.9124
Val Loss: 16828.1909
Epoch 153 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67055.2687
Val Loss: 16559.0176
Epoch 154 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66854.2119
Val Loss: 16559.6230
Epoch 155 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66858.8017
Val Loss: 16594.3252
Epoch 156 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66906.1835
Val Loss: 16558.8190
Epoch 157 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66830.8576
Val Loss: 16706.7901
Epoch 158 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67131.4377
Val Loss: 16547.1356
Epoch 159 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66784.8103
Val Loss: 16540.2256
Epoch 160 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66774.0506
Val Loss: 16543.4942
Epoch 161 of 900


95it [01:07,  1.40it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66983.0827
Val Loss: 16815.7088
Epoch 162 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66940.9949
Val Loss: 16554.3184
Epoch 163 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66776.4168
Val Loss: 16533.8341
Epoch 164 of 900


95it [01:07,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66738.4918
Val Loss: 16559.5911
Epoch 165 of 900


95it [01:06,  1.42it/s]
24it [00:15,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66795.7674
Val Loss: 16555.4438
Epoch 166 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66744.4967
Val Loss: 16524.8688
Epoch 167 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66733.8466
Val Loss: 16553.5709
Epoch 168 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66767.3701
Val Loss: 16564.5992
Epoch 169 of 900


95it [01:06,  1.44it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66813.1062
Val Loss: 16518.9582
Epoch 170 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66682.2939
Val Loss: 16521.1427
Epoch 171 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66689.1494
Val Loss: 16520.2075
Epoch 172 of 900


95it [01:05,  1.45it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66683.4438
Val Loss: 16554.6126
Epoch 173 of 900


95it [01:05,  1.46it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66926.5728
Val Loss: 16540.0222
Epoch 174 of 900


95it [01:05,  1.45it/s]
24it [00:15,  1.51it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66688.5986
Val Loss: 16551.6482
Epoch 175 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66674.1658
Val Loss: 16634.4434
Epoch 176 of 900


95it [01:05,  1.44it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66726.6459
Val Loss: 16518.2970
Epoch 177 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66846.3589
Val Loss: 16519.7730
Epoch 178 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66622.7923
Val Loss: 16504.1248
Epoch 179 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66605.3332
Val Loss: 16493.8461
Epoch 180 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66597.2286
Val Loss: 16493.3801
Epoch 181 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.50it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66648.0061
Val Loss: 16510.7793
Epoch 182 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66591.8732
Val Loss: 16547.4379
Epoch 183 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66636.5887
Val Loss: 16487.1802
Epoch 184 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66569.1091
Val Loss: 16482.9894
Epoch 185 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66578.5439
Val Loss: 16568.3846
Epoch 186 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66587.9096
Val Loss: 16484.2292
Epoch 187 of 900


95it [01:07,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66560.3851
Val Loss: 16496.8959
Epoch 188 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66556.3830
Val Loss: 16620.2058
Epoch 189 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 67031.4145
Val Loss: 16524.4859
Epoch 190 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66565.7736
Val Loss: 16538.9114
Epoch 191 of 900


95it [01:08,  1.38it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66577.3879
Val Loss: 16472.9714
Epoch 192 of 900


95it [01:08,  1.39it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66513.0031
Val Loss: 16466.3396
Epoch 193 of 900


95it [01:08,  1.39it/s]
24it [00:16,  1.42it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66486.9835
Val Loss: 16481.2919
Epoch 194 of 900


95it [01:08,  1.38it/s]
24it [00:16,  1.43it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66485.2754
Val Loss: 16461.8197
Epoch 195 of 900


95it [01:08,  1.38it/s]
24it [00:16,  1.42it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66509.9166
Val Loss: 16489.0048
Epoch 196 of 900


95it [01:08,  1.39it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66467.9471
Val Loss: 16488.6069
Epoch 197 of 900


95it [01:06,  1.42it/s]
24it [00:16,  1.49it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66492.4546
Val Loss: 16460.2257
Epoch 198 of 900


95it [01:07,  1.40it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66454.4630
Val Loss: 16471.2222
Epoch 199 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66480.7618
Val Loss: 16452.4191
Epoch 200 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66439.2782
Val Loss: 16605.5505
Epoch 201 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66583.2490
Val Loss: 16478.0354
Epoch 202 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66420.6444
Val Loss: 16484.0550
Epoch 203 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66418.8653
Val Loss: 16461.5823
Epoch 204 of 900


95it [01:07,  1.40it/s]
24it [00:16,  1.45it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66446.4731
Val Loss: 16469.0075
Epoch 205 of 900


95it [01:07,  1.40it/s]
24it [00:16,  1.44it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66438.5459
Val Loss: 16441.8323
Epoch 206 of 900


95it [01:07,  1.40it/s]
24it [00:16,  1.48it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66394.5721
Val Loss: 16560.0985
Epoch 207 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66435.3433
Val Loss: 16436.4211
Epoch 208 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66343.0156
Val Loss: 16561.6290
Epoch 209 of 900


95it [01:07,  1.41it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66504.0635
Val Loss: 16432.6708
Epoch 210 of 900


95it [01:06,  1.43it/s]
24it [00:16,  1.46it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66355.7568
Val Loss: 16433.9502
Epoch 211 of 900


95it [01:07,  1.42it/s]
24it [00:16,  1.47it/s]
  0%|          | 0/94 [00:00<?, ?it/s]

Train Loss: 66321.1367
Val Loss: 16439.7507
Epoch 212 of 900


 71%|███████▏  | 67/94 [00:47<00:18,  1.42it/s]

In [1]:
import os 

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.realpath('__file__')))
print(PROJECT_PATH)
#DATAPATH = os.path.join(PROJECT_PATH,'content/drive/MyDrive','maps_data')
DATAPATH = os.path.join(PROJECT_PATH,'content/drive/MyDrive','MapsDataCities')
print(DATAPATH)
MODEL_PATH = os.path.join(PROJECT_PATH,'content/drive/MyDrive/TrainedModels','ModelBeta_900e_15k_lr001')

/
/content/drive/MyDrive/MapsDataCities


In [2]:
import torch
torch.save(model.state_dict(), MODEL_PATH)

NameError: ignored

In [ ]:
model_loaded = ConvVAE().to(device)
model_loaded.load_state_dict(torch.load(MODEL_PATH))

<All keys matched successfully>

In [ ]:
test_device = "cpu"
model_loaded.to(test_device)

ConvVAE(
  (enc1): Conv2d(3, 8, kernel_size=(4, 4), stride=(2, 2), padding=(4, 4))
  (enc2): Conv2d(8, 16, kernel_size=(4, 4), stride=(2, 2))
  (enc3): Conv2d(16, 32, kernel_size=(4, 4), stride=(2, 2))
  (enc4): Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2))
  (enc5): Conv2d(64, 8, kernel_size=(4, 4), stride=(2, 2))
  (dec1): ConvTranspose2d(8, 64, kernel_size=(4, 4), stride=(2, 2))
  (dec2): ConvTranspose2d(64, 32, kernel_size=(4, 4), stride=(2, 2))
  (dec3): ConvTranspose2d(32, 16, kernel_size=(4, 4), stride=(2, 2))
  (dec4): ConvTranspose2d(16, 8, kernel_size=(4, 4), stride=(2, 2))
  (dec5): ConvTranspose2d(8, 3, kernel_size=(4, 4), stride=(2, 2))
)

In [ ]:
import numpy as np
#z_images = np.zeros((,392))
image_names = []
image_tensors = []
with torch.no_grad():
    for idx, image in enumerate(train_loader):
        #print(image.shape)
        #z_image = model_loaded.encode(image)[0]#.numpy()
        print(train_loader.dataset.__getname__(idx))
        print(idx)
        #print(torch.flatten(z_image, start_dim=1).shape)
        #print(z_image)
        #z_images[idx] = z_image
        #print(z_images[idx].shape)
        #image_names.append(image_name)
        #image_tensors.append(image)

**INTERPOLATION EXPERIMENT**

In [ ]:
idx1 = 129
idx2 = 250
#requires batch_size 1!!!
def interpolate(idx1,idx2):
  make_dir()
  npoints = 30
  points = []
  gif_images = []
  with torch.no_grad():
      z1 = model_loaded.encode(train_loader.dataset[idx1].unsqueeze(0))[0].numpy()
      z2 = model_loaded.encode(train_loader.dataset[idx2].unsqueeze(0))[0].numpy()
      print(train_loader.dataset.__getname__(idx1))
      print(train_loader.dataset.__getname__(idx2))
    
      # Interpolation
      diff = z2-z1
      for i in range(0,npoints+1):
          point = z1 + i*diff/npoints
          points.append(point)
        
      # Plot each interpolated point
      for i,point in enumerate(points):
          image_tensor = model_loaded.decode(torch.from_numpy(point))
          gif_image = make_grid(image_tensor.cpu())
          gif_images.append(gif_image)
      image_to_vid(gif_images)

In [ ]:
interpolate(10, 120)


torch.Size([1, 3, 286, 286])
torch.Size([1, 3, 286, 286])
19752.png
20817.png


In [ ]:
w_out = ( - 1) * 2 - 2 * 0 + 1 * (10-1) + 0 + 1
print(w_out)

W_out=(W_in−1)×stride[1]−2×padding[1]+dilation[1]×(kernel_size[1]−1) +output_padding[1]+1 302 - 148 -